datasource/ethereum/src/block_stream.rs

```diff
@@ -499,59 +494,78 @@ where
-                                        subgraph_ptr,
-                                        descendant_parent_ptr,
-                                    ).unwrap();
-                            }
+                    Box::new(descendant_blocks.map(move |descendant_block| {
+                        // First, check if there are blocks between subgraph_ptr and
+                        // descendant_block.
+                        let descendant_parent_ptr =
+                            EthereumBlockPointer::to_parent(&descendant_block);
+                        if subgraph_ptr != descendant_parent_ptr {
+                            // descendant_block is not a direct child.
+                            // Therefore, there are blocks that are irrelevant to this subgraph
+                            // that we can skip.
+
+                            debug!(
+                                ctx.logger,
+                                "Skipping {} block(s) with no relevant events...",
+                                descendant_parent_ptr.number - subgraph_ptr.number
+                            );
 
-                            // Update our copy of the subgraph ptr to reflect the
-                            // value it will have after descendant_block is
-                            // processed.
-                            subgraph_ptr = (&descendant_block).into();
+                            // Update subgraph_ptr in store to skip the irrelevant blocks.
+                            ctx.subgraph_store
+                                .set_block_ptr_with_no_changes(
+                                    ctx.subgraph_id.clone(),
+                                    subgraph_ptr,
+                                    descendant_parent_ptr,
+                                ).unwrap();
+                        }
 
-                            descendant_block
-                        },
-                    )) as Box<Stream<Item = _, Error = _> + Send>,
+                        // Update our copy of the subgraph ptr to reflect the
+                        // value it will have after descendant_block is
+                        // processed.
+                        subgraph_ptr = (&descendant_block).into();
+
+                        descendant_block
+                    })) as Box<Stream<Item = _, Error = _> + Send>,
                 )))
             }
         }
     }
 
+    /// Load Ethereum blocks in bulk, returning results as they come back as a Stream.
+    fn load_blocks(
+        &self,
+        block_hashes: Vec<H256>,
+    ) -> impl Stream<Item = EthereumBlock, Error = Error> + Send {
+        let ctx = self.clone();
+
+        let block_hashes_batches = block_hashes
+            .chunks(200) // max batch size
+            .map(|chunk| chunk.to_vec())
+            .collect::<Vec<_>>();
+
+        // Return a stream that lazily loads batches of blocks
+        stream::iter_ok::<_, Error>(block_hashes_batches)
```

Comment:
The batching + chunking logic here is pretty darn elegant.

Reply:
Thanks :blush: